reshuffles names for blackbox/whitebox contexts, changes bundle notation

xeno-by opened the pull request:
review @retronym

xeno-by committed:
*boxContext => *box.Context	, *boxMacro => *box.Macro

Performs the following renamings:
* scala.reflect.macros.BlackboxContext to scala.reflect.macros.blackbox.Context
* scala.reflect.macros.BlackboxMacro to scala.reflect.macros.blackbox.Macro
* scala.reflect.macros.WhiteboxContext to scala.reflect.macros.whitebox.Context
* scala.reflect.macros.WhiteboxMacro to scala.reflect.macros.whitebox.Macro

https://groups.google.com/forum/#!topic/scala-internals/MX40-dM28rk

xeno-by committed:
changes bundles to be classes, not traits extending Macro

Adjusts bundle notation to read `class Bundle(val c: Context)` instead of
`class Bundle extends Macro`. This avoids calling compileLate in the
macro compiler and associated tooling problems.

retronym commented on src/compiler/scala/reflect/macros/runtime/JavaReflectionRuntimes.scala:
```diff
@@ -19,7 +19,7 @@ trait JavaReflectionRuntimes {
       args => {
         val implObj =
-          if (isBundle) implClass.getConstructor(classOf[ApiContext]).newInstance(args.c)
+          if (isBundle) implClass.getConstructors().head.newInstance(args.c)
```
Doesn't handle overloading. For simplicity, you might want to rule out references to macro bundles with overloaded constructors.

xeno-by replied:
Nice catch. Overloading is not a problem as long as we have a constructor of a suitable type. Will fix this.

xeno-by replied:
Also, the macro bundle checker looks only at the primary constructor, and we can't have overloaded primary constructors :)

xeno-by committed:
prohibits constructor overloading for macro bundles

As per Jason’s feedback, this commit handles overloaded constructors
in macro bundles. The backend now checks that we have a constructor of
a correct type. The frontend now prohibits multiple constructors altogether.

xeno-by committed:
removes Scala reflection-based macro runtime

This code is currently unused, so I’m removing it in order to avoid bitrot.
Once we’re sure that we can trust Scala reflection with our macros, the
removed file can be easily reinstated.

retronym commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -603,32 +599,31 @@ trait Definitions extends api.StandardDefinitions {
+      val ctor = tp.erasure.typeSymbol.primaryConstructor
+      ctor.paramss match {
+        case List(List(c)) =>
+          if (c.info != ErrorType && c.info != NoType && c.info != NothingTpe && c.info != NullTpe) NoType
```
discarded `if`

xeno-by replied:
very embarassing, thank you!

retronym commented:
PLS SYNCH

scala-jenkins commented:
(kitty-note-to-self: ignore 32245966)
:cat: Synchronaising! :pray:

retronym commented:
LGTM. The naming is consistent with the collections library. Getting rid of `compileLate` is massive.

xeno-by commented:
Can we merge this into master or not yet?

retronym commented:
Yes, I think this one makes sense to merge pre-M8. Doing so now.

retronym commented:
Do you support macro bundles in 2.10.x via paradise? If so, it would be great to publish a new version of that to coincide with M8.

xeno-by commented:
@retronym No, I don't, because they needed `*Macro` classes. But now I very well could. Good idea, thanks!
